### ISSUE TITLE
opencv and

### ISSUE THREAD

**Issue by ljjcoder** [icoz69/DeepEMD#3]

**Comment by ljjcoder** [comment]
@icoz69 Thanks for your excellent work! I have a question: Will the use of opencv solver affect the accuracy during training? Because you mentioned in the paper that the training process uses qpth solver.

**Comment by icoz69** [comment]
hi. yes, it will, which i have stated in the readme file. 
usually it cause around 1 % performance drop, as the gradient can not flow back through the constraints.

**Comment by ljjcoder** [comment]
@icoz69 Thanks for your reply.

**Comment by ljjcoder** [comment]
@icoz69  Sorry to bother you again.  When I train DeepEMD-FCN with a pre-trained model, I encountered the following problem.
![1592932966(1)](https://user-images.githubusercontent.com/40750189/85434867-40440c00-b5b9-11ea-81d6-2ab06a6e3132.png)
Can you give me some suggestions to solve it?

**Comment by icoz69** [comment]
hello. it seems the issue happens in the qpth solver. I would suggest you update all dependency to the latest, including pytorch, qpth, cvxpy...

**Comment by ljjcoder** [comment]
@icoz69 Sorry to bother you again. I updated the related libraries and the above problem was solved. However, When I use 4GPUs to train the code(python train_pretrain.py -dataset miniimagenet -gpu 0,1,2,3), it encountered new problems：
![1592990766(1)](https://user-images.githubusercontent.com/40750189/85529614-d4f14d00-b63f-11ea-9892-d51125f6b952.png)
When I switch to 3 GPUs for training(python train_pretrain.py -dataset miniimagenet -gpu 0,1,2), I can train normally.
![1592991022(1)](https://user-images.githubusercontent.com/40750189/85530297-68c31900-b640-11ea-95d0-5a2137799936.png)
I would like to ask, are there any special requirements for the number of GPUs in order to train the second stage of normal training? In order to reproduce your results, how many cards should I use for training?

**Comment by ljjcoder** [comment]
@icoz69 sorry, the command should be 

python train_meta.py -deepemd fcn -shot 1 -way 5 -solver qpth -gpu 0,1,2

**Comment by icoz69** [comment]
hi. these is no special requirements for the number of GPUs. I can use 1~4 gpus to train the model. Can you please print the shape of the input tensors to the emd_inference_qpth function, i.e. the similarity_map, weight1, weight2 ?

**Comment by ljjcoder** [comment]
@icoz69 hi，after I upgraded pytorch to 1.5.1, the above error did not appear. But there are new errors like this:
![1593659588(1)](https://user-images.githubusercontent.com/40750189/86312336-13e15d00-bc55-11ea-9ddd-81114b6f8607.png)
After I changed the querysize from 15 to 16, the program can run normally again:
![1593659871(1)](https://user-images.githubusercontent.com/40750189/86312606-b0a3fa80-bc55-11ea-9177-2a471397b1dd.png)
  I tested the case where the number of queries is 8, 12 and they all work normally. So I guess that the number of queries must be an integer multiple of the number of GPUs used.

**Comment by icoz69** [comment]
thanks for the observation. i guess the problem happens due to the multi-gpu learning.

**Comment by ljjcoder** [comment]
@icoz69 sorry to bother you again. After the meta stage training is over, the accuracy of the model will be automatically tested. However, we found that the accuracy rate obtained by train_meta.py will be higher than the accuracy rate obtained by eval.py (I checked the code and they are all testing 5000 tasks on the test data set). Is there any difference between them?

**Comment by icoz69** [comment]
hello. this is due to the randomness in this task. in other words, the sampled data for evaluation are different  in two scripts. You can enlarge  the number of episodes to achieve a more stable result, but as far as I have seen, the gap is already pretty small for 5000 episodes.
